### PR TITLE
Stable softmax cross entropy

### LINKED ISSUE #584: Stable SoftmaxCrossEntropy
The probabilities from the softmax function can become numerically zero. Hence,  for the forward method in the softmax_cross_entropy something like

 	numpy.log(numpy.maximum(1.0e-8,p))

and

	log(fmax(1.0E-8, y[_j * n_channel + t]) ) 

should be added for the cpu and gpu, respectively.

### PR DESCRIPTION
**Pull request by unnonouno**
fix #584 
I wrote a code which uses `CUDNN_SOFTMAX_LOG`, but cudnn v2 doesn't support it.

### CODE REVIEW

**Comment by delta2323** [comment]
> I wrote a code which uses CUDNN_SOFTMAX_LOG, but cudnn v2 doesn't support it.
So you mean you give up using cudnn, don't you?

**Comment by unnonouno** [comment]
In this PR, I didn't include fix for cudnn because we need a new API in cupy. I'll make a new issue.

**Comment by delta2323** [comment]
OK, then I will check this PR alone.

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 1]
```diff
@@ -48,14 +62,15 @@ def forward_cpu(self, inputs):
         if count == 0:
             return numpy.zeros((), dtype=x.dtype),
 
-        y = (numpy.log(p) * (t.flat != self.ignore_label)).sum(keepdims=True) \
+        y = (log_p * (t.flat != self.ignore_label)).sum(keepdims=True) \
             * (-1.0 / count)
```
Comment (okuta) [inline chainer/functions/loss/softmax_cross_entropy.py]: Use `force_type` because `(-1.0 / count)` returns the value which is numpy.float64 type.

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 2]
```diff
@@ -65,12 +80,12 @@ def forward_gpu(self, inputs):
...
-        y = cupy.rollaxis(self.y, 1, self.y.ndim)
+        log_y = cupy.rollaxis(log_y, 1, log_y.ndim)
         ret = cuda.reduce(
-            'S t, raw T y, int32 n_channel, T inv_count', 'T out',
-            't == -1 ? 0 : log(y[_j * n_channel + t])',
+            'S t, raw T log_y, int32 n_channel, T inv_count', 'T out',
+            't == -1 ? 0 : log_y[_j * n_channel + t]',
             'a + b', 'out = a * inv_count', '0', 'crossent_fwd'
-        )(t, y.reduced_view(), y.shape[-1], -1.0 / count)
+        )(t, log_y.reduced_view(), log_y.shape[-1], -1.0 / count)
```
Comment (okuta) [inline chainer/functions/loss/softmax_cross_entropy.py]: Use `force_type`.

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 3]
```diff
@@ -32,11 +45,12 @@ def check_type_forward(self, in_types):
...
-        self.y, = softmax.Softmax().forward((x,))
-        yd = numpy.rollaxis(self.y, 1)
-        yd = yd.reshape(len(yd), -1).T
+        log_y = softmax_log(x)
+        self.y = numpy.exp(log_y)
+        log_yd = numpy.rollaxis(log_y, 1)
+        log_yd = log_yd.reshape(len(log_yd), -1).T
 
-        p = yd[six.moves.range(t.size), numpy.maximum(t.flat, 0)]
+        log_p = log_yd[six.moves.range(t.size), numpy.maximum(t.flat, 0)]
```
Comment (okuta) [inline chainer/functions/loss/softmax_cross_entropy.py]:
```python
log_yd = log_yd.reshape(len(log_yd), -1)
log_p = log_yd[numpy.maximum(t.flat, 0), six.moves.range(t.size)]
```
is better.

**Comment by okuta** [comment]
LGTM except commnets!

File: chainer/functions/loss/softmax_cross_entropy.py (hunk 1 above)
Comment (unnonouno) [inline chainer/functions/loss/softmax_cross_entropy.py]: "`ndarray of float32` * `float64`" is `ndarray of float32`, isn't it?

Comment (okuta) [inline chainer/functions/loss/softmax_cross_entropy.py]: I think that type of `count` might be `ndarray`.
https://github.com/pfnet/chainer/pull/712#discussion-diff-46940915R57
What do you think?

Comment (unnonouno) [inline chainer/functions/loss/softmax_cross_entropy.py]: I tried to raise an error with `count == 65536`, but it causes no error.
 `count` is scalar, isn't it?

Comment (okuta) [inline chainer/functions/loss/softmax_cross_entropy.py]: Sorry, I made a mistake.

**Comment by okuta** [comment]
LGTM!
